subsys/modem/modem_cmux.c

```diff
@@ -1261,6 +1296,11 @@ int modem_cmux_attach(struct modem_cmux *cmux, struct modem_pipe *pipe)
 	ring_buf_reset(&cmux->transmit_rb);
 	cmux->receive_state = MODEM_CMUX_RECEIVE_STATE_SOF;
 	modem_pipe_attach(cmux->pipe, modem_cmux_bus_callback, cmux);
+
+	K_SPINLOCK(&cmux->work_lock) {
```

Comment:
The modem_cmux_attach() and modem_cmux_release() are only performed by the thread which owns the cmux instance, so there is no scenario where modem_cmux_attach() is called simultaneously to modem_cmux_release()

Reply:
Okay then if the assumption is that a single thread will attach/detach (should that be documented?).

> and since the instance is released

It could actually be attached already as nothing checks that the instance is released...

> The spinlock is here as the modem_pipe_attach() may trigger work

So shouldn't we put `modem_pipe_attach()` inside the spinlock to really be safe?

Reply:
> Okay then if the assumption is that a single thread will attach/detach (should that be documented?).

This is assumed for all of the CMUX APIs, only the "owner" calls connect, attach, release, init etc. other threads only use DLCI channels through modem pipes. Do you have a good way to word this in the docs?

> It could actually be attached already as nothing checks that the instance is released...

Added check to ensure it is indeed released

> So shouldn't we put modem_pipe_attach() inside the spinlock to really be safe?

It really is not needed, adding code to a critical/spinlock section increases the likelihood of re-locking the spinlock, causing a hard fault. The spinlock is only for preventing start of work, and modifying the `attached` var

Reply:
> This is assumed for all of the CMUX APIs, only the "owner" calls connect, attach, release, init etc. other threads only use DLCI channels through modem pipes. Do you have a good way to word this in the docs?

Something like:
`.. warning:: The CMUX instances are not protected against concurrent access. It is assumed that only a single thread calls ``modem_cmux_*()`` functions at the same time on a same CMUX instance.`

here:
https://github.com/zephyrproject-rtos/zephyr/blob/7f8cc43a0b5e675569866b7364ab34ba8ebd4e0a/doc/services/modem/index.rst?plain=1#L41
?